File: src/types/events.ts
/**
 * TypeScript payload shapes for Tauri events emitted by the Rust backend
 * (see `src-tauri/src/events/mod.rs`). Nothing emits these yet in M1 — the
 * backend doesn't exist — but `src/lib/events.ts` is typed against these
 * contracts so M2+ wiring is a matter of implementation, not design.
 */

import type {
  ActivityEvent,
  AgentRunStatus,
  AgentStatus,
  MissionStatus,
  Notification,
  TaskStatus,
  ToolCall,
} from "./db";

export interface AgentStatusChangedEvent {
  agentId: string;
  status: AgentStatus;
}

export interface AgentRunStatusChangedEvent {
  agentRunId: string;
  agentId: string;
  status: AgentRunStatus;
}

export interface AgentRunActivityEvent {
  agentRunId: string;
  event: ActivityEvent;
}

export interface ToolCallUpdatedEvent {
  agentRunId: string;
  toolCall: ToolCall;
}

/** Ephemeral streamed assistant text — never persisted; a `model_message`
 * `AgentRunActivityEvent` carries the final text once the turn completes. */
export interface AgentRunMessageDeltaEvent {
  agentRunId: string;
  text: string;
}

export interface NotificationCreatedEvent {
  notification: Notification;
}

/** M9: a mission-level status transition, emitted by `orchestrator::scheduler::run_mission`. */
export interface MissionStatusChangedEvent {
  missionId: string;
  status: MissionStatus;
}

/**
 * M9: one task belonging to a running mission changed status. `reason` is
 * only set when the scheduler itself determined *why* (a dependency cycle,
 * or a failed upstream dependency) — it isn't persisted on the `tasks` row,
 * just surfaced live, so it should be treated as ephemeral (gone on reload).
 */
export interface MissionTaskUpdatedEvent {
  missionId: string;
  taskId: string;
  status: TaskStatus;
  reason: string | null;
}

export interface TerminalOutputEvent {
  terminalId: string;
  chunk: string;
}

export interface TerminalExitEvent {
  terminalId: string;
  exitCode: number | null;
}

/** Event-name -> payload-type map, used by `src/lib/events.ts`'s typed `listen()` wrapper. */
export interface ForgeEventMap {
  "agent:status-changed": AgentStatusChangedEvent;
  "agent-run:status-changed": AgentRunStatusChangedEvent;
  "agent-run:activity": AgentRunActivityEvent;
  "agent-run:tool-call-updated": ToolCallUpdatedEvent;
  "agent-run:message-delta": AgentRunMessageDeltaEvent;
  "notification:created": NotificationCreatedEvent;
  "mission:status-changed": MissionStatusChangedEvent;
  "mission:task-updated": MissionTaskUpdatedEvent;
  "terminal:output": TerminalOutputEvent;
  "terminal:exit": TerminalExitEvent;
}

export type ForgeEventName = keyof ForgeEventMap;

File: src-tauri/src/orchestrator/events.rs
//! Typed event-emission helpers for M9's scheduler, on the same pattern as
//! `agent::events` — kept in one place so `scheduler.rs` and
//! `commands::mission_commands` agree on exactly one payload shape per
//! event, each mirroring the corresponding type in `src/types/events.ts`.

use serde::Serialize;
use tauri::AppHandle;

use crate::db::models::{MissionStatus, TaskStatus};
use crate::error::AppResult;
use crate::events;

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct MissionStatusChangedPayload<'a> {
    mission_id: &'a str,
    status: MissionStatus,
}

/// Mirrors `MissionStatusChangedEvent` in `src/types/events.ts`.
pub fn mission_status_changed(app: &AppHandle, mission_id: &str, status: MissionStatus) -> AppResult<()> {
    events::emit(app, events::MISSION_STATUS_CHANGED, MissionStatusChangedPayload { mission_id, status })
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct MissionTaskUpdatedPayload<'a> {
    mission_id: &'a str,
    task_id: &'a str,
    status: TaskStatus,
    /// Set only when the scheduler itself determined *why* (e.g. a
    /// dependency cycle, or a failed upstream dependency) — not persisted on
    /// the `tasks` row itself (no schema column for it), just surfaced live
    /// so the UI can show an honest reason without inventing one.
    reason: Option<&'a str>,
}

/// Mirrors `MissionTaskUpdatedEvent` in `src/types/events.ts`.
pub fn mission_task_updated(
    app: &AppHandle,
    mission_id: &str,
    task_id: &str,
    status: TaskStatus,
    reason: Option<&str>,
) -> AppResult<()> {
    events::emit(app, events::MISSION_TASK_UPDATED, MissionTaskUpdatedPayload { mission_id, task_id, status, reason })
}
